consensus/reactor.go

```diff
@@ -858,31 +1005,67 @@ func (ps *PeerState) ApplyCommitStepMessage(msg *CommitStepMessage) {
 	ps.ProposalBlockParts = msg.BlockParts
 }
 
-func (ps *PeerState) ApplyHasVoteMessage(msg *HasVoteMessage) {
+func (ps *PeerState) ApplyProposalPOLMessage(msg *ProposalPOLMessage) {
 	ps.mtx.Lock()
 	defer ps.mtx.Unlock()
 
 	if ps.Height != msg.Height {
 		return
 	}
+	if ps.ProposalPOLRound != msg.ProposalPOLRound {
+		return
+	}
 
-	ps.setHasVote(msg.Height, msg.Round, msg.Type, msg.Index)
+	// TODO: Merge onto existing ps.ProposalPOL?
+	// We might have sent some prevotes in the meantime.
+	ps.ProposalPOL = msg.ProposalPOL
 }
 
-func (ps *PeerState) ApplyProposalPOLMessage(msg *ProposalPOLMessage) {
+func (ps *PeerState) ApplyHasVoteMessage(msg *HasVoteMessage) {
 	ps.mtx.Lock()
 	defer ps.mtx.Unlock()
 
 	if ps.Height != msg.Height {
 		return
 	}
-	if ps.ProposalPOLRound != msg.ProposalPOLRound {
-		return
+
+	ps.setHasVote(msg.Height, msg.Round, msg.Type, msg.Index)
+}
+
+// When a peer claims to have a maj23 for some BlockID at H,R,S,
+// we will try to respond with a VoteSetBitsMessage showing which
+// bits we already have (and which we don't yet have),
+// but that happens in another goroutine.
+func (ps *PeerState) ApplyVoteSetMaj23Message(msg *VoteSetMaj23Message) {
+	// ps.mtx.Lock()
+	// defer ps.mtx.Unlock()
+
+	select {
+	case ps.Maj23Queue <- msg:
+	default:
+		// Just ignore if we're already processing messages.
 	}
+}
 
-	// TODO: Merge onto existing ps.ProposalPOL?
-	// We might have sent some prevotes in the meantime.
-	ps.ProposalPOL = msg.ProposalPOL
+// The peer has responded with a bitarray of votes that it has
+// of the corresponding BlockID.
+// ourVotes: BitArray of votes we have for msg.BlockID
+// NOTE: if ourVotes is nil (e.g. msg.Height < rs.Height),
+// we conservatively overwrite ps's votes w/ msg.Votes.
+func (ps *PeerState) ApplyVoteSetBitsMessage(msg *VoteSetBitsMessage, ourVotes *BitArray) {
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
+	votes := ps.getVoteBitArray(msg.Height, msg.Round, msg.Type)
+	if votes != nil {
+		if ourVotes == nil {
+			votes.Update(msg.Votes)
+		} else {
+			otherVotes := votes.Sub(ourVotes)
```

Comment:
I don't understand. We don't know what blockID these `votes` are for - shouldn't we always overwrite with `msg.Votes` (which is for a known blockID)? Otherwise, to do the subtraction (ie. find votes the peer has that we don't), wouldn't we have to track blockID in the `ps`?